include/openrave/kinbody.h

```diff
@@ -3066,6 +3083,25 @@ class OPENRAVE_API KinBody : public InterfaceBase
     ConfigurationSpecification _spec;
     CollisionCheckerBasePtr _selfcollisionchecker; ///< optional checker to use for self-collisions
 
+public:
+    // Gary registers forward kinematics functions
+    struct ForwardKinematicsStruct {
+        ForwardKinematicsStruct();
+        ModuleBaseConstPtr pCalculatorModule = nullptr; ///< kinbody basic calculators module
+        boost::function<bool(const std::vector<double>&)> pSetLinkTransformsFn; ///< function that sets links' transforms
+        boost::function<void(std::vector<double>&)> pGetDOFLastSetValuesFn; ///< function that updates kinbody's dof values
+        bool bInitialized = false; ///< indicator of successful initialization
+    };
+
+    /// \brief Associate the kinbody's current kinematics geometry hash with a forward kinematics structure
+    /// \param[in] fkstruct a forward kinematics structure
+    /// \param[in] bOverWrite whether to replace a registered forward kinematics structure if it exists
+    ///
+    /// \return true if the forward kinematics structure to be registered has all the necessary facilitities we can use. 
+    bool RegisterForwardKinematicsStruct(const ForwardKinematicsStruct& fkstruct, const bool bOverWrite=false);
+
+protected:
+    std::map<std::string, ForwardKinematicsStruct> _mHash2ForwardKinematicsStruct; ///< maps a kinematics geometry hash to a forward kinematics structure
```

Comment:
@tgn3000 checking for a hash in the map every call is not very efficient. i think there should be a way for you to get a `ForwardKinematicsStruct _currentFK` variable that gets initialized from your DB of `ForwardKinematicsStruct` during the compute internal initialization. That way, you never have to check anything before using it.

Reply:
Now I use `_pCurrentForwardKinematicsStruct` to cache the currently used  `ForwardKinematicsStruct` with the correct kinematics geometry hash.